### PR TITLE
feat(repository): add findByForeignKeys function

### PR DESCRIPTION
**Pull request by nabdelgadir**
Implemented initial version of the new helper function `findByForeignKeys`. 

Resolves https://github.com/strongloop/loopback-next/issues/3443.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/repository/src/relations/relation.helpers.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+export async function findByForeignKeys<
+  Target extends Entity,
+  TargetID,
+  TargetRelations extends object,
+  ForeignKey
+>(
+  targetRepository: EntityCrudRepository<Target, TargetID, TargetRelations>,
+  fkName: StringKeyOf<Target>,
+  fkValues: ForeignKey[],
```
Comment (raymondfeng) [inline packages/repository/src/relations/relation.helpers.ts]: Can we accept a single value too? For example:

```ts
fkValues: ForeignKey[] | ForeignKey,
```
```

File: packages/repository/src/relations/relation.helpers.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * @param targetRepository - The target repository where the model instances are found
+ * @param fkName - Name of the foreign key
+ * @param fkValues - Array of the values of the foreign keys to be included
+ * @param scope - Additional scope constraints (not currently supported)
+ * @param options - Options for the operations
+ */
+export async function findByForeignKeys<
+  Target extends Entity,
+  TargetID,
```
Comment (raymondfeng) [inline packages/repository/src/relations/relation.helpers.ts]: Do we really need this? Use `unknown` is probably good enough.

File: packages/repository/src/relations/relation.helpers.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/repository/src/relations/relation.helpers.ts]: Is it possible to have a type param such as `FK extends StringKeyOf<Target>` and use it to constrain fk values, such as `Target[FK]`?

File: packages/repository/src/relations/relation.helpers.ts (hunk 2 above)
Comment (nabdelgadir) [reply packages/repository/src/relations/relation.helpers.ts]: `TargetID`? but we use it in [`targetRepository: EntityCrudRepository<Target, TargetID, TargetRelations>,`](https://github.com/strongloop/loopback-next/pull/3473/files#diff-791ac353d48a6890b8c35dfaa1c836a1R24)

Comment (raymondfeng) [reply packages/repository/src/relations/relation.helpers.ts]: We can remove `TargetID` from the type params and use `unknown`:

 targetRepository: EntityCrudRepository<Target, unknown, TargetRelations>,

Comment (nabdelgadir) [reply packages/repository/src/relations/relation.helpers.ts]: Ohh I see, thanks! What's the benefit of using `unknown`?

Comment (raymondfeng) [reply packages/repository/src/relations/relation.helpers.ts]: The `TargetID` does not matter in our usage of `find` as no `findById` is involved.

File: packages/repository/src/relations/relation.helpers.ts (hunk 1 above)
Comment (agnes512) [reply packages/repository/src/relations/relation.helpers.ts]: Let me reword it to make sure that I understand the type constrains in `tsc` correctly:

if we do `foreignKey extends StringKeyOf<Target>`, it would constrain the passed in `foreignKey` to be a string-like `<Target>` instance?

Comment (raymondfeng) [reply packages/repository/src/relations/relation.helpers.ts]: I'm proposing the following:

```ts
export async function findByForeignKeys<
  Target extends Entity,
  FK extends StringKeyOf<Target>,
  TargetRelations extends object,
>(
  targetRepository: EntityCrudRepository<Target, unknown, TargetRelations>,
  fkName: FK,
  fkValues: Target[FK][],
  scope?: Filter<Target>,
  options?: Options,
): Promise<(Target & TargetRelations)[]> {
}
```

File: packages/repository/src/relations/relation.helpers.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  options?: Options,
+): Promise<(Target & TargetRelations)[]> {
+  // throw error if scope is defined and non-empty
+  // see https://github.com/strongloop/loopback-next/issues/3453
+  if (scope && !_.isEmpty(scope)) {
+    throw new Error('scope is not supported');
+  }
+
+  const value = Array.isArray(fkValues)
+    ? fkValues.length === 1
```
Comment (raymondfeng) [inline packages/repository/src/relations/relation.helpers.ts]: Let's add a shortcut for empty array to return `[]` immediately. Include a test to cover this corner case too.

File: packages/repository/src/relations/relation.helpers.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      : {inq: fkValues}
+    : fkValues;
+
+  const where = ({[fkName]: value} as unknown) as Where<Target>;
+  const targetFilter = {where};
+
+  return targetRepository.find(targetFilter, options);
+}
+
+export type StringKeyOf<T> = Extract<keyof T, string>;
```
Comment (raymondfeng) [inline packages/repository/src/relations/relation.helpers.ts]: Maybe it's not necessary to export this type?

File: packages/repository/src/relations/relation.helpers.ts (hunk 3 above)
Comment (nabdelgadir) [reply packages/repository/src/relations/relation.helpers.ts]: If we pass in an empty array, it complains that `Type 'never[]' is not assignable to type 'number'.` (or whatever the type of the foreign key value is)

Comment (raymondfeng) [reply packages/repository/src/relations/relation.helpers.ts]: You can do the following:
```ts
const fkIds: number[] = []; 
```
